crates/wit-bindgen-go/src/interface.rs

```diff
@@ -3769,34 +3753,81 @@ func (v *{name}) WriteToIndex(w {wrpc}.ByteWriter) (func({wrpc}.IndexWriter) err
+            if ty.flags.len() % 8 > 0 {
+                buf_len += 1;
+            }
+
             uwriteln!(
                 self.src,
-                "func (v *{name}) WriteToIndex(w {wrpc}.ByteWriter) (func({wrpc}.IndexWriter) error, error) {{"
+                r#"func (v *{name}) WriteToIndex(w {wrpc}.ByteWriter) (func({wrpc}.IndexWriter) error, error) {{
+                    var p [{buf_len}]byte
+                    var b byte
+"#
             );
-            self.push_str("var n ");
-            self.int_repr(repr);
-            self.push_str("\n");
+
             for (i, Flag { name, .. }) in ty.flags.iter().enumerate() {
+                let shift = i % 8;
                 self.push_str("if v.");
                 self.push_str(&name.to_upper_camel_case());
-                self.push_str(" {\n");
-                if i <= 64 {
-                    uwriteln!(self.src, "n |= 1 << {i}");
-                } else {
-                    let errors = self.deps.errors();
-                    uwriteln!(
-                        self.src,
-                        r#"return nil, {errors}.New("encoding `{name}` flag value would overflow 64-bit integer, flags containing more than 64 members are not supported yet")"#
-                    );
+                uwriteln!(
+                    self.src,
+                    r#"{{
+        b |= 1 << {shift}                
+    }}
+                    "#,
+                );
+                if shift == 7 || i == ty.flags.len() - 1 {
+                    let offset = i / 8;
+                    uwriteln!(self.src, "p[{offset}] = b\n\tb = 0");
+                }
+            }
+            uwriteln!(
+                self.src,
+                r#"
+    _, err := w.Write(p[:])
+    return nil, err
+}}
+                "#,
+            );
+
+            // ReadFromIndex()
+            uwrite!(
+                self.src,
+                r#"func (v *{name}) ReadFromIndex(r {io}.Reader) error {{
+        var p [{buf_len}]byte
+        if _, err := r.Read(p[:]); err != nil {{
+            return err
+        }}
+    
+        var b byte
+    "#
+            );
+
+            for (i, Flag { name, .. }) in ty.flags.iter().enumerate() {
+                if i % 8 == 0 {
+                    let offset = i / 8;
+                    self.push_str("\n");
+                    uwriteln!(self.src, r#"b = p[{offset}]"#);
                 }
+
+                self.push_str("if b % 2 == 1 {\n");
+                self.push_str("v.");
+                self.push_str(&name.to_upper_camel_case());
+                self.push_str(" = true\n");
                 self.push_str("}\n");
+
+                if i != ty.flags.len() - 1 {
+                    self.push_str("b >>= 1\n\n");
+                }
             }
-            self.push_str("return nil, ");
-            self.print_write_discriminant(repr, "n", "w");
-            self.push_str("\n");
-            self.push_str("}\n");
```

Comment:
for cases when `ty.flags.len() % 8 > 0`, we should should also check that `p[{}] <= 1 << {}` for `ty.flags.len()/8`, `ty.flags.len()%8`, i.e. ensure that there are no trailing bits set that we don't know what to do with

Reply:
I don't think this is needed. The value is set to zero value initially, which is simply `0`. So I don't see how the bits could be set.

Reply:
Basically, what I'm trying to address here is the possibility of receiving invalid data, e.g. if a peer sends us `0b1000` for a type `flags {a, b, c}`, then we should fail parsing, since we don't know what to do with that 4th bit, otherwise we'd accept and invalid encoding of a value

Reply:
40dcbffbdc946d2d4d35a4af3ee3e68139d2f368